File: 백준/Gold/10026. 적록색약/적록색약.cc
#include <iostream>

using namespace std;
int nx[4] = {1,0,-1,0};
int ny[4] = {0,1,0,-1};
int save[100][100];
bool visited[100][100];
void bfs(int x, int y);
int x, y, n, normal, blindness;
char c;
//R은 1, B는 2, G는 3
int main() {
    cin >> n;
    for(int i=0; i<n; i++){
        for(int j=0; j<n; j++){
            cin >> c;
            if(c=='R') save[i][j] = 1;
            else if(c=='B') save[i][j] = 2;
            else if(c=='G') save[i][j] = 3;
        }
    }
    
    for(int i=0 ;i<n; i++){
        for(int j=0; j<n; j++){
            if(!visited[i][j]) {
                normal++;
                bfs(i,j);
            }
        }
    }
    
    for(int i=0; i<n; i++){
        for(int j=0; j<n; j++){
            visited[i][j] = false;
            if(save[i][j] == 3) save[i][j] = 1;
        } 
    }
    
    for(int i=0 ;i<n; i++){
        for(int j=0; j<n; j++){
            if(!visited[i][j]) {
                blindness++;
                bfs(i,j);
            }
        }
    }
    cout << normal <<" "<<blindness<<'\n';
    return 0;
}
void bfs(int x, int y){
    visited[x][y] = true;
    for(int i=0; i<4; i++){
        int dx = x + nx[i];
        int dy = y + ny[i];
        
        if(dx<0 || dy<0 || dx>=n || dy>=n) continue;
        if(visited[dx][dy]) continue;
        if(save[x][y] != save[dx][dy]) continue;
        bfs(dx, dy);
    }
}

File: 백준/Gold/10026. 적록색약/적록색약.java
import java.util.*;
import java.io.*;

class Main {
    public static int n;
    public static char[][] grid;
    public static boolean[][] visited;
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        n = Integer.parseInt(br.readLine());
        grid = new char[n][n];
        visited = new boolean[n][n];

        for(int i=0; i<n; i++){
            String line = br.readLine();
            for(int j=0; j<n; j++){
                grid[i][j] = line.charAt(j);
            }
        }

        int normal = bfs();

        visited = new boolean[n][n];
        for(int i=0; i<n; i++){
            for(int j=0; j<n; j++){
                if(grid[i][j] == 'R') grid[i][j] = 'G';
            }
        }

        int blindness = bfs();
        
        System.out.println(normal + " " + blindness);
        
    }
    public static int bfs(){
        Queue<int[]> q = new LinkedList<>();
        int dx[] = {0,0,1,-1};
        int dy[] = {-1,1,0,0};
        int area = 0;

        for(int i=0; i<n; i++){
            for(int j=0; j<n; j++){
                if(visited[i][j]) continue;
                visited[i][j] = true;
                area++;
                q.add(new int[]{i,j});

                while(!q.isEmpty()){
                    int[] curr = q.poll();
                    int y = curr[0];
                    int x = curr[1];

                    for(int dir=0; dir<4; dir++){
                        int ny = y + dy[dir];
                        int nx = x + dx[dir];

                        if(ny < 0 || nx < 0 || ny >= n || nx >= n) continue;
                        if(visited[ny][nx] || grid[y][x] != grid[ny][nx]) continue;
                        visited[ny][nx] = true;
                        q.add(new int[]{ny, nx});
                    }
                }
            }
        }
        return area;
    }
}